integration-tests/ahm/src/mock.rs

```diff
@@ -33,39 +31,80 @@ use polkadot_primitives::UpwardMessage;
 use polkadot_runtime::{
 	Block as PolkadotBlock, RcMigrator, Runtime as Polkadot, RuntimeEvent as RcRuntimeEvent,
 };
-use remote_externalities::{Builder, Mode, OfflineConfig, RemoteExternalities};
+use remote_externalities::{Builder, Mode, OfflineConfig};
 use runtime_parachains::{
 	dmp::DownwardMessageQueues,
 	inclusion::{AggregateMessageOrigin as RcMessageOrigin, UmpQueueId},
 };
+use sp_io::TestExternalities;
 use sp_runtime::{BoundedVec, Perbill};
 use std::str::FromStr;
+use tokio::sync::Mutex as TokioMutex;
 use xcm::prelude::*;
 //use frame_support::traits::QueueFootprintQuery; // Only on westend
 
 pub const AH_PARA_ID: ParaId = ParaId::new(1000);
 const LOG_RC: &str = "runtime::relay";
 const LOG_AH: &str = "runtime::asset-hub";
 
+pub enum Chain {
+	Relay,
+	AssetHub,
+}
+
+impl ToString for Chain {
+	fn to_string(&self) -> String {
+		match self {
+			Chain::Relay => "SNAP_RC".to_string(),
+			Chain::AssetHub => "SNAP_AH".to_string(),
+		}
+	}
+}
+
+pub type Snapshot = (Vec<(Vec<u8>, (Vec<u8>, i32))>, sp_core::H256);
+
+static RC_CACHE: TokioMutex<Option<Snapshot>> = TokioMutex::const_new(None);
+static AH_CACHE: TokioMutex<Option<Snapshot>> = TokioMutex::const_new(None);
+
 /// Load Relay and AH externalities in parallel.
-pub async fn load_externalities(
-) -> Option<(RemoteExternalities<PolkadotBlock>, RemoteExternalities<AssetHubBlock>)> {
+pub async fn load_externalities() -> Option<(TestExternalities, TestExternalities)> {
 	let (rc, ah) = tokio::try_join!(
-		tokio::spawn(async { remote_ext_test_setup::<PolkadotBlock>("SNAP_RC").await }),
-		tokio::spawn(async { remote_ext_test_setup::<AssetHubBlock>("SNAP_AH").await })
+		tokio::spawn(async { remote_ext_test_setup(Chain::Relay).await }),
+		tokio::spawn(async { remote_ext_test_setup(Chain::AssetHub).await })
 	)
 	.ok()?;
 	Some((rc?, ah?))
 }
 
-pub async fn remote_ext_test_setup<Block: sp_runtime::traits::Block>(
-	env: &str,
-) -> Option<RemoteExternalities<Block>> {
+pub async fn remote_ext_test_setup(chain: Chain) -> Option<TestExternalities> {
 	sp_tracing::try_init_simple();
-	let snap = std::env::var(env).ok()?;
+	log::info!("Checking {} snapshot cache", chain.to_string());
+
+	// Check cache first
+
+	// Hold the lock for the entire function to avoid loading the snapshot twice.
+
+	let mut cache = match chain {
+		Chain::Relay => RC_CACHE.lock().await,
+		Chain::AssetHub => AH_CACHE.lock().await,
+	};
+
+	if let Some(snapshot) = cache.as_ref().cloned() {
+		log::info!("Using cached snapshot for {}", chain.to_string());
+		return Some(TestExternalities::from_raw_snapshot(
+			snapshot.0,
+			snapshot.1,
+			sp_storage::StateVersion::V1,
+		));
+	}
+
+	log::info!("Loading {} snapshot", chain.to_string());
+
+	// Load snapshot.
+	let snap = std::env::var(chain.to_string()).ok()?;
 	let abs = std::path::absolute(snap.clone());
 
-	let ext = Builder::<Block>::default()
+	let ext = Builder::<PolkadotBlock>::default()
```

Comment:
it requires it to access the hasher function. we know it is the same for RC and AH